File: dsa_learning_series/week2/5_ZCO12001.py
in = int(input())
s = list(map(int, input().split()))

stack = 0
nesting_d = 0
nesting_p = 1

max_l = 0
len_p = 1
previous = 0
for i in range(n):
    if s[i]==1:
        stack+=1
    else:
        if stack>nesting_d:
            nesting_d = stack
            nesting_p = i
        stack-=1
    if stack==0:
        l = i-previous+1
        if l>max_l:
            max_l = l
            len_p = previous
        previous = i+1

print(nesting_d, nesting_p, max_l, len_p+1)

File: dsa_learning_series/week2/5_ZCO12001.cpp
#include<iostream>
#include<vector>
#include<algorithm>

using namespace std;

int main(){
	int n;
	cin >> n;
	vector<int> s;
	for(int i=0;i<n;i++){
		int inp;
		cin >> inp;
		s.push_back(inp);
	};

	int stack = 0;
	int nesting_p = 1;
	int nesting_l = 0;
	

	int max_l = 0;
	int len_p = 1;
	int previous = 0;

	for(int i=0;i<n;i++){
		if (s[i]==1){
			stack++;
		}else{	
			if(stack>nesting_l){
				nesting_l = stack;
				nesting_p = i;
			}
			stack--;
		}
		if(stack==0){
			int l = i-previous+1;
			if(l>max_l){
				max_l = l;
				len_p = previous;
			}
			previous = i+1;
		}
	};

	cout << nesting_l << " " << nesting_p <<  " " << max_l << " " << len_p+1 << endl;
	return 0;
}
